src/executor/evaluate/mod.rs

```diff
@@ -197,6 +199,72 @@ pub async fn evaluate<'a, T: 'static + Debug>(
         Expr::Wildcard | Expr::QualifiedWildcard(_) => {
             Err(EvaluateError::UnreachableWildcardExpr.into())
         }
+        Expr::Case {
+            operand,
+            when_then,
+            else_result,
+        } => {
+            let operand = match operand {
+                Some(expr) => Some(eval(expr).await?),
+                None => None,
+            };
+
+            let when_then: Vec<(Evaluated, Evaluated)> = stream::iter(when_then.iter())
+                .map(Ok::<_, Error>)
+                .and_then(|(when, then)| async move { Ok((eval(when).await?, eval(then).await?)) })
+                .try_collect::<Vec<_>>()
+                .await?;
+
+            let else_result = match else_result {
+                Some(result) => Some(eval(result).await?),
+                None => None,
+            };
+
+            let (whens, thens) = when_then
+                .clone()
+                .into_iter()
+                .unzip::<_, _, Vec<_>, Vec<_>>();
+
+            let results = match &else_result {
+                Some(er) => [thens.clone(), [er.to_owned()].to_vec()].concat(),
+                None => thens.clone(),
+            }
+            .into_iter()
+            .map(|result| result.try_into())
+            .collect::<Result<Vec<Value>>>()?;
+
+            if !results
+                .iter()
+                .map(|result| discriminant(result))
+                .all_equal()
+            {
+                Err(EvaluateError::UnequalResultTypes("CASE".to_owned()).into())
+            } else {
+                match operand {
```

Comment:
`operand` is only being used in here, not above. So, you don't need to split `operand` handling codes into two different positions.

You can also use `find_map` here to merge `find` and `map_or`.

```rust
match operand {                                                 
    Some(op) => {                                               
        let op = eval(op).await?;                               
                                                                
        Ok(when_then                                            
            .into_iter()                                        
            .find_map(|(when, then)| when.eq(&op).then(|| then))
            .unwrap_or_else(|| Evaluated::from(Value::Null)))           
    }    
```

Reply:
Ahah thanks! `find_map` looks very useful :D